Can't download external (private) repositories

Hi,

We've got an issue with downloading external terraform modules on private github namespace.

```
checkov -d .    --quiet   --output github_failed_only   --download-external-modules true    --framework terraform  
Error: -30 14:11:36,9[26](https://github.com/***namespace***/module-tf-cleo-admin/actions/runs/3159383417/jobs/5142549378#step:4:27) [MainThread  ] [ERROR]  failed to get git::https://github.com/***namespace***/module-tf-acm?ref=v4.0.0 because of Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --depth=1 -b v4.0.0 https://github.com/***namespace***/module-tf-acm /github/workspace/.external_modules/github.com/***namespace***/module-tf-acm/v4.0.0
  stderr: 'Cloning into '/github/workspace/.external_modules/github.com/***namespace***/module-tf-acm/v4.0.0'...
fatal: could not read Username for 'https://github.com': No such device or address
'
Error: -30 14:11:38,546 [MainThread  ] [ERROR]  failed to get git::https://github.com/***namespace***/module-tf-wafv2?ref=v3.0.0 because of Cmd('git') failed due to: exit code(1[28](https://github.com/***namespace***/module-tf-cleo-admin/actions/runs/3159383417/jobs/5142549378#step:4:29))
  cmdline: git clone -v --depth=1 -b v3.0.0 https://github.com/***namespace***/module-tf-wafv2 /github/workspace/.external_modules/github.com/***namespace***/module-tf-wafv2/v3.0.0
  stderr: 'Cloning into '/github/workspace/.external_modules/github.com/***namespace***/module-tf-wafv2/v3.0.0'...
fatal: could not read Username for 'https://github.com/': No such device or address
'
```

We tried with `github_pat` variable without success and also tried "git config" hack to replace all `https://github.com` occurences:
```
      # Workaround to allow `terraform init` cloning private repositories
      - run: git config --global url."https://oauth2:${PRIVATE_GITHUB_TOKEN}@github.com".insteadOf "https://github.com"
        env:
          PRIVATE_GITHUB_TOKEN: ${{ secrets.PRIVATE_GITHUB_TOKEN }}
```

Here is our github action configuration for checkov:
```
  static-analysis:
    runs-on: ubuntu-latest
    name: static-analysis
    steps:
      - name: Checkout repo
        uses: actions/checkout@master

      - name: Run Checkov action
        id: checkov
        uses: bridgecrewio/checkov-action@master
        with:
          quiet: true
          soft_fail: false
          framework: terraform
          output_format: github_failed_only
          download_external_modules: true # download external terraform modules from public git repositories and terraform registry
          github_pat: ${{ secrets.PRIVATE_GITHUB_TOKEN }}
          container_user: 1000
```

Can you help ?
If you need some details, feel free to ask.

Thanks !

Same issue here. We use a GitHub App to get tokens for the private GH repos for all of our external module dependencies and then we use the .gitconfig to insert the tokens which are specific to each module/repo.

This works perfectly for Terraform, but not for Checkov which for some reason doesn't respect the .gitconfig settings.

``` 
      - name: Read Required Repos File
        id: get_required_repos
        env:
          REQUIRED_REPOS_FILE_PATH: required_repos.txt
        run: |
          echo "Reading ${REQUIRED_REPOS_FILE_PATH}"
          if [ ! -e "${REQUIRED_REPOS_FILE_PATH}" ]; then echo "${REQUIRED_REPOS_FILE_PATH} not found" && exit 1; fi
          export REQUIRED_REPOSITORIES=$(cat "${REQUIRED_REPOS_FILE_PATH}"| sed 's/\s\+/\n/g' | tr '\n' ',')
          echo "required_repos=$(echo $REQUIRED_REPOSITORIES)" >> $GITHUB_OUTPUT || exit 1
      - name: Generate App tokens
        id: generate_tokens
        uses: OrgNameHere/it_infra_github_app_token@v2
        with:
          app_id: ${{ secrets.IT_INFRA_READ_APP_APPID }}
          private_key: ${{ secrets.IT_INFRA_READ_APP_PRIVATEKEY }}
          # Optional (defaults to ID of the repository's installation).
          # installation_id: 1337
          # Optional (defaults to the current repository).
          repository_names_list:  ${{ steps.get_required_repos.outputs.required_repos }}
      - name: Git Config for Pulling private cross-repo TF modules
        env:
          TF_MODULES_TOKEN_DICT: ${{ steps.generate_tokens.outputs.token_dict }}
        run: |
          for module_name in $(echo "${TF_MODULES_TOKEN_DICT}"|jq -r 'keys[]')
          do
            export THIS_MODULE_TOKEN=$(echo "${TF_MODULES_TOKEN_DICT}"|jq -r .${module_name})
            echo "THIS_MODULE_TOKEN=${THIS_MODULE_TOKEN}"
            git config --global url."https://x-access-token:${THIS_MODULE_TOKEN}@github.com/OrgNameHere/${module_name}.git".insteadOf "https://github.com/OrgNameHere/${module_name}.git"
            # git config --global credential.helper store
          done
``` 

Checkov Action always fails like so:

``` 
BC_REPOSITORY_URL=https://github.com/OrgNameHere/it_infra_terraform_oii_application_sso
running checkov on directory: examples
checkov -d examples       --output sarif  --soft-fail  --download-external-modules true    --framework terraform   
Error: -[27](https://github.com/OrgNameHere/it_infra_terraform_oii_application_sso/actions/runs/4025510047/jobs/6918819710#step:7:28) 14:45:41,478 [MainThread  ] [ERROR]  failed to get git::https://github.com/OrgNameHere/it_infra_terraform_environment_settings?ref=main because of Cmd('git') failed due to: exit code(1[28](https://github.com/OrgNameHere/it_infra_terraform_oii_application_sso/actions/runs/4025510047/jobs/6918819710#step:7:29))
  cmdline: git clone -v --depth=1 -b main -- https://github.com/OrgNameHere/it_infra_terraform_environment_settings /github/workspace/examples/.external_modules/github.com/OrgNameHere/it_infra_terraform_environment_settings/main
  stderr: 'Cloning into '/github/workspace/examples/.external_modules/github.com/OrgNameHere/it_infra_terraform_environment_settings/main'...
fatal: could not read Username for 'https://github.com/': No such device or address
'
``` 

Terraform picks up the insteadOf from the .gitconfig and pulls the modules properly:

``` 
Initializing modules...
Downloading git::https://github.com/OrgNameHere/it_infra_terraform_environment_settings.git?ref=main for oii...
- oii in .terraform/modules/oii
- sso_application in ..
```

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -27,3 +27,5 @@ jobs:
         with:
           directory: tests/private
           soft_fail: true  # don't want to fail the step, because of found findings
+          download_external_modules: true
+          github_pat: ${{ secrets.TEST_PAT_TOKEN }}

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -21,3 +21,9 @@ jobs:
         with:
           directory: terragoat
           soft_fail: true  # don't want to fail the step, because of found findings
+
+      - name: Scan private module
+        uses: ./
+        with:
+          directory: tests/private
+          soft_fail: true  # don't want to fail the step, because of found findings

--- a/README.md
+++ b/README.md
@@ -130,3 +130,29 @@ jobs:
 
 Note that this example uses the latest version (`master`) but you could also use a static version (e.g. `v3`).
 Also, the check ids specified for '--check' and '--skip-check' must be mutually exclusive.
+
+## Example usage for private Terraform modules
+
+To give `checkov` the possibility to download private GitHub modules you need to pass a valid GitHub PAT with the needed permissions.
+
+```yaml
+on: [push]
+jobs:
+  checkov-job:
+    runs-on: ubuntu-latest
+    name: checkov-action
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@master
+
+      - name: Run Checkov action
+        id: checkov
+        uses: bridgecrewio/checkov-action@master
+        with:
+          directory: .
+          soft_fail: true
+          download_external_modules: true
+          github_pat: ${{ secrets.GH_PAT }}
+        env:
+          GITHUB_OVERRIDE_URL: true  # optional: this can be used to instruct the action to override the global GIT config to inject the PAT to the URL
+```

--- a/tests/private/main.tf
+++ b/tests/private/main.tf
@@ -1,0 +1,6 @@
+module "s3_bucket" {
+  source = "github.com/ckv-tests/terraform-aws-s3-bucket-private?ref=v0.0.1"
+
+  acl     = "public"
+  enabled = true
+}

